### DISCUSSION TITLE
Learning Parlai

### DISCUSSION THREAD

**Discussion by MikeyBeez** [facebookresearch/ParlAI#3366] [Q&A] [answered]
Who is having success with display_data.py?  I'm not really sure how to use it.  I get the following:  

>> python display_data.py

Traceback (most recent call last):
  File "display_data.py", line 26, in <module>
    world = create_task(opt, agent)
  File "/home/bard/ParlAI/parlai/core/worlds.py", line 1250, in create_task
    'No task specified. Please select a task with ' + '--task {task_name}.'
RuntimeError: No task specified. Please select a task with --task {task_name}.

**Comment by stephenroller** [comment, marked as answer]
Try `parlai display_data -t convai2` or `python display_data.py -t convai2`. The former is our "super command" which can call all the separate scripts, and is the main directed way.

For a list of all the tasks we have, check out [this page](https://parl.ai/docs/tasks.html).

If you want a good tutorial on using ParlAI, you can go through either our [Colab Tutorial](https://colab.research.google.com/drive/1bRMvN0lGXaTF5fuTidgvlAl-Lb41F7AD#scrollTo=KtVz5dCUmFkN), or through our [Quickstart Guide](https://parl.ai/docs/tutorial_quick.html). If you don't have access to GPUs on your machine, I recommend the colab tutorial, as Google provides free GPUs you can use.

Our full docs are available at https://parl.ai/docs/, though some information is scattered in READMEs through out the repo.

**Reply by MikeyBeez** [reply]
Thank you, Stephen!  I'm so interested in this tech.  I have great hopes for our future.

**Comment by MikeyBeez** [comment]
This is so cool!

**Reply by MikeyBeez** [reply]
I installed apex within my conda parlai environment by cloning the apex repo into the parlai directory and running: 

pip install -v --disable-pip-version-check --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./

When I load the first model in your tutorial I get:

You set --fp16 true with --fp16-impl apex, but fp16 with apex is unavailable. To use apex fp16, please install APEX from https://github.com/NVIDIA/apex.

Should I be able to use apex on my single workstation, or is it for clusters?

**Reply by stephenroller** [reply]
You can use apex on a single workstation. Did you run that command inside the apex folder?

you can use `--fp16-impl mem_efficient` if you don't want/get apex working.

**Comment by MikeyBeez** [comment]
No I didn't run it there.  I'll try moving my jupyter notebook to the apex folder.  I'm guessing this should speed up my training.  BTW, thank you so very much for taking so much time with me.  I really appreciate it.

**Comment by MikeyBeez** [comment]
I made a video about Parlai:  https://www.youtube.com/watch?v=gv3jQrQ_ook

**Reply by stephenroller** [reply]
cool! thanks for sharing!

**Reply by klshuster** [reply]
this is great, thanks for making the video!

**Reply by MikeyBeez** [reply]
I hope to make the Wizard if Wikipedia voice activated tomorrow.  I've already done this with transfer-learning-conv-ai.  Here's a video on how:  https://www.youtube.com/watch?v=1gT2FwwCHBY&t=166s   I need to do more work on the socket.  I think I will need to make it a two way connection.  Thank you for your interest.  It's heartening.  Very few people watch my videos.

**Reply by stephenroller** [reply]
We noticed you found some of the short option names were opaque in the tutorials, so we're revising those to be all long form names for clarity.

We would sincerely welcome any and all additional feedback you have. It's very difficult for us to put ourselves in the shoes of a person seeing these things for the first time.

**Reply by MikeyBeez** [reply]
I'm so happy to assist.  BTW, I'm hoping to make the chat interface natural.  I think I need to modify the local_human.py file to listen on a socket using vosk and reply using gTTS.  Is this the correct way?  Has this been done before?   Is there a switch in interactive.py I can use to call a new speak_and_hear function in local_human.py?

**Reply by MikeyBeez** [reply]
Here's the code I'll need to modify and integrate into local_human.py, if indeed that's the right place to hook:

from gtts import gTTS
import socket

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(("localhost", 9988))
s.listen(1)

while True:
         conn, addr = s.accept()
         data = conn.recv(1024)
        encoding = 'utf-8'
        data = str(data, encoding)
        conn.close()
        #raw_text = input(">>> ")
       #while not raw_text:
       #    print('Prompt should not be empty!')
       #    raw_text = input(">>> ")
       raw_text = data
       history.append(tokenizer.encode(raw_text))
       with torch.no_grad():
           out_ids = sample_sequence(personality, history, tokenizer, model, args)
       history.append(out_ids)
       history = history[-(2*args.max_history+1):]
       out_text = tokenizer.decode(out_ids, skip_special_tokens=True)
       print(out_text)
       text_to_speech = gTTS(text=out_text, lang='en-uk')
       text_to_speech.save('audio.mp3')
      os.system('mpg123 -q audio.mp3')
       #os.system(echo "out_text" | gtts-cli --lang en-gb --nocheck - | mpg123 -)

**Reply by MikeyBeez** [reply]
From my assistant I'll use code from npm:

#############################################################
# STT SPEECH TO TEXT FUNCTION THAT RETURNS THE VARIABLE "chatter"
import pyaudio
import json
from vosk import Model, KaldiRecognizer
#, SetLogLevel

#SetLogLevel(-10)


def myChat():
    # "listens for chatter"
    # We imported vosk up above.
    p = pyaudio.PyAudio()
    # MyFormat = pyaudio.paInt16
    # MyChannels = 1
    # MyRate = 16000
    # MyInput = True
    # MyFPB = 8000
    stream = p.open(format=pyaudio.paInt16, channels=1, rate=16000, input=True, frames_per_buf
fer=8000)
    stream.start_stream()
    model = Model("model")
    rec = KaldiRecognizer(model, 16000)
    while True:
        data = stream.read(2000)
        if len(data) == 0:
            break
        if rec.AcceptWaveform(data):
            myResult = rec.Result()
            myList = myResult.split("text")
            chatter = myList[1]
            stream.stop_stream()
            stream.close()
            p.terminate()
            return chatter
# END STT SPEECH TO TEXT FUNCTION THAT RETURNS THE VARIABLE "chatter"
#############################################################

**Reply by stephenroller** [reply]
https://github.com/facebookresearch/ParlAI/blob/master/parlai/scripts/interactive_web.py#L174-L178 see our interactive_web script for an example of interacting with a model in a webapp. That may be a better example for how to plug in a TTS and STT system

**Reply by MikeyBeez** [reply]
Thank you so much, Stephan.  I want to do this the right way, not just a hack, so I can share it with others.

**Comment by MikeyBeez** [comment]
I'm able to run parlai interactive_web -mf "zoo:tutorial_transformer_generator/model" with no problems.  However parlai interactive_web --model tfidf_retriever -mf zoo:wikipedia_full/tfidf_retriever/model  gives me the following:

parlai interactive_web --model tfidf_retriever -mf zoo:wikipedia_full/tfidf_retriever/model

```
21:55:34 | Overriding opt["model_file"] to /home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model (previously: wiki_full_notitle)
21:55:34 | WARNING: Neither the specified dict file (test_ret.dict) nor the `model_file`.dict file (/home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model.dict) exists, check to make sure either is correct. This may manifest as a shape mismatch later on.
21:55:34 | Loading /home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model.tfidf
21:55:59 | Opt:
21:55:59 |     allow_missing_init_opts: False
21:55:59 |     batchsize: 1
21:55:59 |     datapath: /home/bard/ParlAI/data
21:55:59 |     datatype: train:ordered
21:55:59 |     dict_build_first: True
21:55:59 |     dict_class: None
21:55:59 |     dict_endtoken: __END__
21:55:59 |     dict_file: /home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model.dict
21:55:59 |     dict_include_test: False
21:55:59 |     dict_include_valid: False
21:55:59 |     dict_initpath: None
21:55:59 |     dict_language: english
21:55:59 |     dict_lower: False
21:55:59 |     dict_max_ngram_size: -1
21:55:59 |     dict_maxexs: -1
21:55:59 |     dict_maxtokens: -1
21:55:59 |     dict_minfreq: 0
21:55:59 |     dict_nulltoken: __NULL__
21:55:59 |     dict_starttoken: __START__
21:55:59 |     dict_tokenizer: split
21:55:59 |     dict_unktoken: __UNK__
21:55:59 |     display_add_fields: 
21:55:59 |     display_examples: False
21:55:59 |     display_prettify: False
21:55:59 |     download_path: None
21:55:59 |     dynamic_batching: None
21:55:59 |     evaltask: None
21:55:59 |     hide_labels: False
21:55:59 |     host: localhost
21:55:59 |     image_cropsize: 224
21:55:59 |     image_mode: raw
21:55:59 |     image_size: 256
21:55:59 |     include_labels: True
21:55:59 |     index_by_int_id: True
21:55:59 |     init_model: None
21:55:59 |     init_opt: None
21:55:59 |     interactive_mode: True
21:55:59 |     interactive_task: True
21:55:59 |     local_human_candidates_file: None
21:55:59 |     log_every_n_secs: 2
21:55:59 |     log_keep_fields: all
21:55:59 |     loglevel: info
21:55:59 |     max_train_time: -1
21:55:59 |     model: tfidf_retriever
21:55:59 |     model_file: /home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model
21:55:59 |     multitask_weights: [1]
21:55:59 |     num_epochs: 1.0
21:55:59 |     numthreads: 1
21:55:59 |     outfile: 
21:55:59 |     override: "{'model': 'tfidf_retriever', 'model_file': '/home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model'}"
21:55:59 |     parlai_home: /private/home/kshuster/ParlAI
21:55:59 |     port: 8080
21:55:59 |     remove_title: True
21:55:59 |     retriever_hashsize: 16777216
21:55:59 |     retriever_mode: keys
21:55:59 |     retriever_ngram: 2
21:55:59 |     retriever_num_retrieved: 5
21:55:59 |     retriever_numworkers: None
21:55:59 |     retriever_task: None
21:55:59 |     retriever_tokenizer: simple
21:55:59 |     save_after_valid: False
21:55:59 |     save_every_n_secs: -1
21:55:59 |     save_format: conversations
21:55:59 |     single_turn: False
21:55:59 |     starttime: Jan13_21-55
21:55:59 |     task: wikipedia:full
21:55:59 |     tfidf_context_length: -1
21:55:59 |     tfidf_include_labels: True
21:55:59 |     validation_cutoff: 1.0
21:55:59 |     validation_every_n_secs: -1
21:55:59 |     validation_max_exs: -1
21:55:59 |     validation_metric: accuracy
21:55:59 |     validation_metric_mode: max
21:55:59 |     validation_patience: 10
21:55:59 |     verbose: False
21:55:59 | Current ParlAI commit: 5104b2b954808ba4d0b92271dea0e771ace2924f
21:56:00 | creating task(s): wikipedia:full
21:56:00 | http://localhost:8080/
127.0.0.1 - - [13/Jan/2021 21:56:09] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [13/Jan/2021 21:56:09] "GET /favicon.ico HTTP/1.1" 202 -
127.0.0.1 - - [13/Jan/2021 21:56:16] "POST /interact HTTP/1.1" 200 -
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 33562)
Traceback (most recent call last):
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/socketserver.py", line 316, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/socketserver.py", line 347, in process_request
    self.finish_request(request, client_address)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/socketserver.py", line 720, in __init__
    self.handle()
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/http/server.py", line 426, in handle
    self.handle_one_request()
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/http/server.py", line 414, in handle_one_request
    method()
  File "/home/bard/ParlAI/parlai/scripts/interactive_web.py", line 202, in do_POST
    json_str = json.dumps(model_response)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/bard/miniconda3/envs/parlai/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type ndarray is not JSON serializable
----------------------------------------
```

**Reply by klshuster** [reply]
This is a bug on our end; putting up a fix now, thanks for catching!

**Reply by klshuster** [reply]
Fix is in https://github.com/facebookresearch/ParlAI/pull/3381

**Comment by MikeyBeez** [comment]
Please forgive my terse replies.  I'm autistic; so I often forget to thank people and show appreciation.  I'll try to do better.

**Reply by stephenroller** [reply]
We're very grateful for your feedback and bug reports. Keep it coming.

**Comment by MikeyBeez** [comment]
Here's another of my videos.  This is on parlai's web chat interface.  https://www.youtube.com/watch?v=CzcCxzjusGk

**Reply by MikeyBeez** [reply]
My goodness, Stephen!  You work all the time.  It's great to have a passion for what you do.

**Comment by MikeyBeez** [comment]
Can you please help with these errors/warnings?  parlai interactive --model tfidf_retriever -mf zoo:wikipedia_full/tfidf_retriever/model
15:31:40 | Overriding opt["model_file"] to /home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model (previously: wiki_full_notitle)
15:31:40 | WARNING: Neither the specified dict file (test_ret.dict) nor the `model_file`.dict file (/home/bard/ParlAI/data/models/wikipedia_full/tfidf_retriever/model.dict) exists, check to make sure either is correct. This may manifest as a shape mismatch later on.
    
BTW, I've made more parlai videos.  They're on my Youtube channel.  https://www.youtube.com/channel/UCMxnQ7gyb4BWKbLW5JWiaJA?view_as=subscriber

**Reply by klshuster** [reply]
You can feel free to ignore those.

The first warning is just indicating that the model file path differs from what is in the `model.opt` file (which makes sense, as the `.opt` file saves the absolute location of the model when it was trained, which will change if someone else, e.g. you, downloads and uses the file).

The second warning is primarily a concern for our neural models; most ParlAI agents use some form of a `DictionaryAgent` for converting text in string form to text in token form, i.e. to relevant input to a model. Generally, when loading a pre-trained model, it is important that the dictionary file (a mapping from text to token) is given as well so that the model sees input the same way it was trained; the "shape mismatch" indicated is noting that a default dictionary may not have the same number of words as the dictionary used with the original trained model, which manifests in an error when loading pre-trained embedding weights.

The TFIDF Retriever handles things slightly differently; it uses it's own, pre-built tokenizers that do not require a dictionary file to be provided.

**Reply by MikeyBeez** [reply]
Thank you so much.  Here's my latest learning parlai video:    https://www.youtube.com/watch?v=1h38ouCw06Q

**Comment by MikeyBeez** [comment]
Can someone please explain why there are two switches for model in this:  parlai train_model --model examples/tra  --task convai2 --model-file /tmp/test --batchsize 32  What is the difference between --model and --model-file?   I know I'm being dense, but I can't figure this out.  I know the model-file will be stored in /tmp.  Is --model the type of model?  The architecture?  And if so, where does tra or its specification reside?  Is it /home/bard/ParlAI/parlai/agents/examples/tra.py ? 
 If that is the case, I wonder why it's an agent.  All of your system is so logical once one sees it correctly.  I'm just having trouble here.  I'm probably stuck by my own means.  Cheers and thanks. :)

**Reply by klshuster** [reply]
Hi there, good question. As you have intuited, `--model` refers to the model implementation, or architecture. Models in ParlAI are wrapped in something called an `Agent`, which handles observing incoming `Message`s, and producing outgoing "acts". All models that are specified via this `--model` flag can be found in the `ParlAI/parlai/agents` directory, or subdirectories where necessary - e.g., `--model examples/tra` is in the `ParlAI/parlai/agents/examples/tra.py`. The `Agent` handles everything that is not "model"-specific - it translates from text to model input token ids; it usually records and handles some metrics for the agent; it handles saving and loading trained models; etc. 

The actual "model" itself is usually found in the agent's directory; depending on its complexity, it may be in the same file as the agent itself, or it may be in a `modules.py` file. Generally, the "model" can be accessed from an `agent` object via `agent.model`. 

The terminology can get a bit tricky, but it's due to the fact that everything that interacts within ParlAI is an `Agent` at it's core; so, "teachers" that serve datasets are also `Agent`s themselves, in that they produce outgoing messages to send to the model by iterating over a dataset, and handle incoming messages to compute their own metrics.

Hope that helps!

**Reply by MikeyBeez** [reply]
Yes this helps enormously.  Inside modules.py I've found the architecture info I was expecting to see somewhere.  Bit by bit, I'm learning your system.  First I want to understand the different models I can use immediately for inference and how the models differ in their use cases.  I want to know how the training data for them differs.  Then I want to know how to train models of various type using both existing and new data.  I think I'm mostly getting there.  Then I want to understand how to use a Parlai created interactive_web socket from within a python program.  I haven't been able to write one yet, and I can't find example code.  Finally I want to know how to extend the python code.  I'm reading through that a bit now, hand copying the code, understanding very little, but it is just a first pass through,  Your tutorial and documentation seem to mix a lot of this together.  I'm not criticizing.  I respect what you've done immensely, and I'm very grateful that you've made all this available to researchers.   I only say this because it might help to have an amateur's point of view.   Cheers!
